// @lcdev/ws-rpc/dist/logging.js
"use strict";
/* eslint-disable no-console */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setLogLevel = exports.logger = exports.LogLevel = void 0;
var LogLevel;
(function (LogLevel) {
    LogLevel["Verbose"] = "verbose";
    LogLevel["Info"] = "info";
    LogLevel["Warn"] = "warn";
    LogLevel["Error"] = "error";
    LogLevel["None"] = "none";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
let logLevel = typeof test !== 'undefined' ? LogLevel.Error : LogLevel.Warn;
exports.logger = {
    setLevel(level) {
        logLevel = level;
    },
    verbose(message) {
        if (logLevel === LogLevel.Verbose) {
            console.log(`[ws-rpc][VERBOSE] ${message}`);
        }
    },
    info(message) {
        if (logLevel === LogLevel.Info || logLevel === LogLevel.Verbose) {
            console.log(`[ws-rpc][INFO] ${message}`);
        }
    },
    warn(message) {
        if (logLevel === LogLevel.Warn || logLevel === LogLevel.Info || logLevel === LogLevel.Verbose) {
            console.warn(`[ws-rpc][WARN] ${message}`);
        }
    },
    error(message) {
        if (logLevel === LogLevel.Error ||
            logLevel === LogLevel.Warn ||
            logLevel === LogLevel.Info ||
            logLevel === LogLevel.Verbose) {
            console.error(`[ws-rpc][ERROR] ${message}`);
        }
    },
};
exports.setLogLevel = (level) => exports.logger.setLevel(level);
