// @lcdev/ts/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function ok(value) {
    return exports.res({ value: value, ok: true });
}
exports.ok = ok;
function err(error) {
    return exports.res({ error: error, ok: false });
}
exports.err = err;
/**
 * Extends a Result with combinators.
 */
exports.res = (result) => (Object.assign(Object.assign({}, result), { raw() {
        if (result.ok) {
            return { ok: true, value: result.value };
        }
        return { ok: false, error: result.error };
    },
    unwrap() {
        if (result.ok) {
            return result.value;
        }
        throw result.error;
    },
    or(other) {
        return exports.res(result.ok ? result : other);
    },
    orElse(other) {
        return exports.res(result.ok ? result : other());
    },
    unwrapOr(value) {
        return exports.res(result.ok ? result : ok(value));
    },
    unwrapOrElse(value) {
        return exports.res(result.ok ? result : ok(value()));
    },
    map(map) {
        return exports.res(result.ok ? ok(map(result.value)) : result);
    },
    mapErr(map) {
        return exports.res(result.ok ? result : err(map(result.error)));
    } }));
async function using(resource, fn) {
    try {
        return await fn(resource);
    }
    finally {
        await resource.dispose();
    }
}
exports.using = using;
